### ISSUE TITLE
DragScrollActivity disables tap actions in the semantics tree

### ISSUE THREAD

**Issue by yjbanov** [flutter/flutter#130162]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. This will help a ton: edit [rendering/binding.dart](https://github.com/flutter/flutter/blob/1fb3687bf8bc2b42317cf071ea9ff269db649b75/packages/flutter/lib/src/rendering/binding.dart#L345) and add `debugDumpSemanticsTree();` as the first line in `_handleSemanticsUpdate`.
2. Run `flutter run -d web-server --web-port=8888` on the code sample provided below.
3. Open Chrome and point it at http://localhost:8888.
4. Open Chrome DevTools and enable the mobile emulation mode (this causes Chrome to emulate taps, like on Android).
5. Long-press on the _text_ next to the checkbox. Do not press the checkbox itself (this is key!)

### Expected results

All printed semantics trees should contain a tap action on the node corresponding to the combination of the checkbox and the text next to it, like this (`SemanticsNode#5` is the interesting one; others were stripped of information for readability):

```
SemanticsNode#0
 └─SemanticsNode#1
   └─SemanticsNode#2
     └─SemanticsNode#3
       └─SemanticsNode#4
         └─SemanticsNode#6
           └─SemanticsNode#5
               Rect.fromLTRB(0.0, 0.0, 390.0, 32.0)
               tags: RenderViewport.twoPane
               actions: tap
               flags: hasCheckedState, isFocused, hasEnabledState, isEnabled,
                 isFocusable
               label: "this is a visible label"
               textDirection: ltr
```

### Actual results

Upon receiving a `pointerdown` event, if the `pointerup` doesn't come soon enough `ScrollableState` initiates a drag activity calling `setIgnorePointer` to true. Importantly the `pointerup` event would still produce a click/tap event and win in the gesture arena. The drag activity starts before the tap is detected.

```
at RenderIgnorePointer.describeSemanticsConfiguration$1
at RenderIgnorePointer.get$_semanticsConfiguration
at RenderIgnorePointer.markNeedsSemanticsUpdate$0
at RenderIgnorePointer.set$ignoring
at ScrollableState.setIgnorePointer$1
at ScrollPositionWithSingleContext.beginActivity$1
at ScrollPositionWithSingleContext.beginActivity$1
at ScrollableState._scrollable$_handleDragStart$1
```

The reason `ScrollPositionWithSingleContext` ignores the pointer is because `DragScrollActivity.shouldIgnorePointer` returns true. If I change the implementation of the getter to return false, the problem goes away.

This causes a semantic update, leading to the following semantics tree being synced into the engine prior to the `pointerup` event:

```
SemanticsNode#0
 └─SemanticsNode#1
   └─SemanticsNode#2
     └─SemanticsNode#3
       └─SemanticsNode#4
         └─SemanticsNode#6
           └─SemanticsNode#5
               Rect.fromLTRB(0.0, 0.0, 390.0, 32.0)
               tags: RenderViewport.twoPane
               flags: hasCheckedState, isFocused, hasEnabledState, isEnabled,
                 isFocusable
               label: "this is a visible label"
               textDirection: ltr
```

Notice that `actions: tap` is gone. This tells the engine that the node is no longer tappable and causes the engine to stop sending tap events to it.

Shortly after, another semantics update reinstates the tap action. However, by then it's too late as the click was ignored.

To summarize, the sequence of events is as follows:

* Initial semantics update adds the tap action to the checkbox
* `pointerdown` happens
* Drag start => ignore pointer injected
* Semantics update removes the tap action
* `pointerup` happens
* `click` happens - because the node no longer has the tap action the click is ignored
* Another semantics update adds the tap action back, but it's too late

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

void main() {
  runApp(const MyApp());
  if (kIsWeb) {
    SemanticsBinding.instance.ensureSemantics();
  }
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Accessibility tester app',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Accessibility tester app'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return const Scaffold(
      body: CheckboxTest(),
    );
  }
}

class CheckboxTest extends StatefulWidget {
  const CheckboxTest({ super.key });

  @override
  State<CheckboxTest> createState() => _CheckboxTestState();
}

class _CheckboxTestState extends State<CheckboxTest> {
  bool _value1 = false;

  @override
  Widget build(BuildContext context) {
    // ListView is necessary for the scrolling infra to kick in.
    return ListView(
      children: <Widget>[
        Row(
          children: [
            Checkbox(
              autofocus: true,
              value: _value1,
              onChanged: (bool? value) {
                setState(() {
                  _value1 = value!;
                });
              },
            ),
            const Text('this is a visible label'),
          ],
        ),
      ],
    );
  }
}
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

https://github.com/flutter/flutter/assets/211513/0744bc2e-c10b-40ad-b022-68c809abd661

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[!] Flutter (Channel [user-branch], 3.12.0-4.0.pre.316, on macOS 13.4.1 22F82 darwin-arm64, locale en)
    ! Flutter version 3.12.0-4.0.pre.316 on channel [user-branch] at /Users/yjbanov/code/flutter/flutter
      Currently on an unknown channel. Run `flutter channel` to switch to an official channel.
      If that doesn't fix the issue, reinstall Flutter by following instructions at
      https://flutter.dev/docs/get-started/install.
    ! Upstream repository unknown source is not a standard remote.
      Set environment variable "FLUTTER_GIT_URL" to unknown source to dismiss this error.
[!] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.
[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.2)
[✓] VS Code (version 1.79.2)
[✓] Connected device (2 available)
[✓] Network resources

! Doctor found issues in 2 categories.
```

</details>

**Comment by goderbauer** [comment]
cc @chunhtai for a11y and @Piinks for scrolling

**Comment by yjbanov** [comment]
Did a little more digging, and check this out:

```
Gesture arena 1    ❙ ★ Opening new gesture arena.
Gesture arena 1    ❙ Adding: VerticalDragGestureRecognizer#cf174(start behavior: start)
Gesture arena 1    ❙ Closing with 1 member.
Gesture arena 1    ❙ Default winner: VerticalDragGestureRecognizer#cf174(start behavior: start)
```

So I think this is not entirely DragScrollActivity's fault. Since the pointer events are landing outside the checkbox that provides a `TapGestureRecognizer` there's no competition in the arena. Only the semantics tree knows that the tappable area is bigger than the checkbox. So `VerticalDragGestureRecognizer` wins by default, and it calls `setIgnorePointer` under the assumption that nothing but a vertical drag can happen at this point.

**Comment by chunhtai** [comment]
>  Since the pointer events are landing outside the checkbox

So in this case the tap should not have registered, and should be WAI?

**Comment by yjbanov** [comment]
The screen reader sends `pointerdown`, `pointerup`, and `click` events in the middle of the DOM node. Since the framework merged the checkbox and the text into one bigger node, the middle of _that_ node is outside the bounds of the checkbox. Since we ignore events landing there, the screen reader isn't able to communicate taps, which is the bug I'm trying to fix.

<img width="865" alt="Screenshot 2023-07-07 at 4 11 30 PM" src="https://github.com/flutter/flutter/assets/211513/50e4f9f2-d36b-4820-93d2-c53714f33b08">

**Comment by yjbanov** [comment]
Currently, there are several things conspiring to prevent taps from reaching the widget:

* The web engine sees `pointerdown` and `pointerup`, and it ignores the `click` because it believes that the framework will figure it out. As you can see from the illustration above, that breaks down easily. However, when I fix that I hit the next issue:
* Gesture arena and `DragScrollActivity` collectively disable the tap action on the semantics node upon `pointerdown`, causing the engine to ignore taps and clicks. I added a workaround for this in the web engine that sends `SemanticsAction.tap` to the framework anyway. The next issue I hit:
* Since the tap action is disabled in the framework, the framework ignores the tap action too.

The only way I can see out of this situation via a pure engine-side fix is to delay events starting from `pointerdown` until the `pointerup` before allowing the framework to do its gesture recognition. This way the gesture arena won't be involved until we know if there's a click on the combined node. The tradeoff is that it will delay reaction time for scrolling until the click is detected or rejected. I'm going to try this approach and see if it still produces usable results.

**Comment by goderbauer** [comment]
> This way the gesture arena won't be involved until we know if there's a click on the combined node. The tradeoff is that it will delay reaction time for scrolling until the click is detected or rejected. I'm going to try this approach and see if it still produces usable results.

If I understand this correctly, the same delay would happen in the framework, though, if an area can register taps and drags. There'd be a small delay while the recognizers battle it out in the gesture arena before we can be confident that this is a drag or tap.

**Comment by yjbanov** [comment]
@goderbauer When there's only one gesture recognizer (as is the case at the center of the combined node), there's no delay in the framework. `VerticalDragGestureRecognizer` wins by default immediately after `pointerdown`. The only way to delay it is for the engine to delay `pointerdown`.

I am testing a solution that seems like it might work. I delay event sequences starting with `pointerdown` for up to 200ms. If within those 200ms I see a click, I send it to the framework as `SemanticsAction.tap`. Otherwise, I send the pointer events to the framework and drop the DOM click event. The tradeoff is that even when the drag gesture detector is the only one present, there's a 200ms delay before dragging can start. However, it seems to be a better trade-off than missing clicks entirely.

**Comment by yjbanov** [comment]
Update: I'm planning to land https://github.com/flutter/engine/pull/43620 to fix the more severe issue of missed clicks. That does come with the trade-off of delaying scrolling by 200ms. The effect would be limited to apps that always enable semantics. Apps with semantics turned off are not affected. The 200ms is also not a dealbreaker when using screen readers, because screen readers do not scroll using gestures. It only happens when you both have semantics on but you do not use the screen reader, which is not a common use-case.

**Comment by yjbanov** [comment]
Closing for now. https://github.com/flutter/engine/pull/43620 seems to be sufficient to address this issue. If we see new reports related to this, we can reevaluate.
